pkg/scheduler/plugins/elasticquota/plugin.go

```diff
@@ -0,0 +1,565 @@
+		elasticQuotaInfos: NewSimpleElasticQuotaInfos(),
+		failOver:          make(chan struct{}),
+	}
+	elasticQuota.groupQuotaManager.SetScaleMinQuotaEnabled(true)
+	elasticQuota.quotaOverUsedRevokeController = NewQuotaOverUsedRevokeController(
+		*elasticQuota.pluginArgs.ContinueOverUseCountTriggerEvict, elasticQuota.groupQuotaManager)
+
+	podInformer := handle.SharedInformerFactory().Core().V1().Pods()
+	podInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    elasticQuota.OnPodAdd,
+			UpdateFunc: elasticQuota.OnPodUpdate,
+			DeleteFunc: elasticQuota.OnPodDelete,
+		},
+	)
+
+	nodeInformer := handle.SharedInformerFactory().Core().V1().Nodes().Informer()
+	nodeInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    elasticQuota.OnNodeAdd,
+			UpdateFunc: elasticQuota.OnNodeUpdate,
+			DeleteFunc: elasticQuota.OnNodeDelete,
+		})
+
+	elasticQuotaInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    elasticQuota.OnQuotaAdd,
+			UpdateFunc: elasticQuota.OnQuotaUpdate,
+			DeleteFunc: elasticQuota.OnQuotaDelete,
+		})
+
+	elasticQuota.elasticQuotaController = NewElasticQuotaController(kubernetes.NewForConfigOrDie(handle.KubeConfig()),
+		elasticQuotaInformer.Lister(), client, elasticQuota.groupQuotaManager)
+	ctx := context.TODO()
+
+	// need to recover quota first
+	scheSharedInformerFactory.Start(ctx.Done())
+	scheSharedInformerFactory.WaitForCacheSync(ctx.Done())
+	close(elasticQuota.failOver)
+
+	handle.SharedInformerFactory().Start(ctx.Done())
+	handle.SharedInformerFactory().WaitForCacheSync(ctx.Done())
+	return elasticQuota, nil
+}
+
+func (g *Plugin) Start() {
+	g.createDefaultQuotaIfNotPresent()
+	go wait.Until(g.RevokePodDueToQuotaOverUsed, RevokePodCycle, nil)
+	go wait.Until(g.elasticQuotaController.Run, time.Second, context.TODO().Done())
+}
+
+func (g *Plugin) Name() string {
+	return Name
+}
+
+//PreFilter performs the following validations.
+// 1  refresh plugin's elasticQuotaInfos before Write state
+// 2. Check if the (pod.request + eq.used) is less than eq.runtime.
+func (g *Plugin) PreFilter(ctx context.Context, state *framework.CycleState, pod *corev1.Pod) *framework.Status {
+	quotaName := g.getPodAssociateQuotaName(pod)
+
+	// update elasticQuotaInfo
+	quotaInfo := g.groupQuotaManager.GetQuotaInfoByName(quotaName)
+	quotaUsed := quotaInfo.GetUsed()
+	quotaRuntime := quotaInfo.GetRuntime()
+	g.elasticQuotaInfos[quotaName].setUsed(quotaUsed)
+	g.elasticQuotaInfos[quotaName].setRuntime(quotaRuntime)
+
+	snapshotElasticQuota := g.snapshotElasticQuota()
+	state.Write(SnapshotStateKey, snapshotElasticQuota)
+
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+	newUsed := quotav1.Add(podRequest, quotaUsed)
+
+	if noOverUse, _ := quotav1.LessThanOrEqual(newUsed, quotaRuntime); !noOverUse {
+		return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("pod:%v  is rejected in PreFilter because"+
```

Comment:
Please improve the readability of the message. e.g "Scheduling refused due to insufficient quotas, quotaName: %v, runtime: %v, used: %v, request: %v". By the way, when return the status with `framework.Unscheduable` in PreFilter stage, then scheduler will exec the preemption.  So is the preemption expected?

Reply:
the former is modified, the latter(is the preemption expected) will discuss with @buptcozy

Reply:
If the current Pending Pod has a higher priority, we should trigger preemption within the same quota. Is it right?

Reply:
yes, after discussion with @buptcozy, preFilter will always return Success, the check if the (pod.request + eq.used) is less than eq.runtime will be done in Filter.

Reply:
after discussion with @buptcozy , we can't guarantee the elastic Quota's PreFilter is the last to run. And if there exist two or more plugins, which claims should be the last plugin to run, how can users organize the order.

Reply:
At present trigger preemption if failed to PreFilter as there is insufficient quota.